contracts/src/v0.8/ccip/test/pools/GHO/fork/GhoTokenPoolMigrate1_4To1_5/TokenPoolsUpgrade.t.sol

```diff
@@ -0,0 +1,123 @@
+
+    vm.expectEmit();
+    emit CCIPSendRequested(_messageToEvent(message, 220, 1, feeTokenAmount, alice, l1.metadataHash, uint32(90000)));
+    vm.prank(alice);
+    l1.router.ccipSend{value: feeTokenAmount}(l2.chainSelector, message);
+  }
+
+  function test_releaseOrMintFrom1_2OffRamp() public {
+    uint256 amount = 10e18;
+    {
+      vm.selectFork(l1.forkId);
+      uint256 balanceBefore = l1.token.balanceOf(alice);
+      // mock release on legacy offramp
+      vm.prank(l1.EVM2EVMOffRamp1_2);
+      l1.tokenPool.releaseOrMint(abi.encode(alice), alice, amount, l2.chainSelector, "");
+      assertEq(l1.token.balanceOf(alice), balanceBefore + amount);
+    }
+    {
+      vm.selectFork(l2.forkId);
+      uint256 balanceBefore = l2.token.balanceOf(alice);
+      // mock release on legacy offramp
+      vm.prank(l2.EVM2EVMOffRamp1_2);
+      l2.tokenPool.releaseOrMint(abi.encode(alice), alice, amount, l1.chainSelector, "");
+      assertEq(l2.token.balanceOf(alice), balanceBefore + amount);
+    }
+  }
+
+  function test_releaseOrMintFrom1_5OffRamp() public {
+    uint256 amount = 10e18;
+    {
+      vm.selectFork(l1.forkId);
+      uint256 balanceBefore = l1.token.balanceOf(alice);
+      // mock release on legacy offramp
+      vm.prank(l1.EVM2EVMOffRamp1_5);
+      l1.tokenPool.releaseOrMint(abi.encode(alice), alice, amount, l2.chainSelector, "");
+      assertEq(l1.token.balanceOf(alice), balanceBefore + amount);
+    }
+    {
+      vm.selectFork(l2.forkId);
+      uint256 balanceBefore = l2.token.balanceOf(alice);
+      // mock release on legacy offramp
+      vm.prank(l2.EVM2EVMOffRamp1_5);
+      l2.tokenPool.releaseOrMint(abi.encode(alice), alice, amount, l1.chainSelector, "");
+      assertEq(l2.token.balanceOf(alice), balanceBefore + amount);
+    }
+  }
+
+  function _upgradeExistingLockReleaseTokenPool() internal {
+    UpgradeableLockReleaseTokenPool_Sepolia poolImpl = new UpgradeableLockReleaseTokenPool_Sepolia(
+      address(l1.token),
+      l1.tokenPool.getArmProxy(),
+      l1.tokenPool.getAllowListEnabled(),
+      l1.tokenPool.canAcceptLiquidity()
+    );
+    _upgradeProxy(TransparentUpgradeableProxy(payable(address(l1.tokenPool))), address(poolImpl));
+  }
+
+  function _upgradeExistingBurnMintTokenPool() internal {
+    UpgradeableBurnMintTokenPool_ArbSepolia poolImpl = new UpgradeableBurnMintTokenPool_ArbSepolia(
+      address(l2.token),
+      l2.tokenPool.getArmProxy(),
+      l2.tokenPool.getAllowListEnabled()
+    );
+    _upgradeProxy(TransparentUpgradeableProxy(payable(address(l2.tokenPool))), address(poolImpl));
+  }
+
+  function _upgradeProxy(TransparentUpgradeableProxy proxy, address impl) private {
+    address proxyAdminAddress = address(
+      uint160(uint256(vm.load(address(proxy), bytes32(uint256(keccak256("eip1967.proxy.admin")) - 1))))
+    );
+    assertNotEq(proxyAdminAddress, address(0), "version mismatch: proxyAdmin");
+    if (proxyAdminAddress.code.length != 0) {
```

Comment:
is there any reason you prefer not to prank the proxy admin if it's a contract?

Reply:
i wrote these functions such that they're sort of reusable in the deploy script, ideally if the upgrade scripts were also in this repo then i would've created the deploy script and this test inherit & re-use the upgrade proxy & setPoolProxy methods to test the deploy script as well

Reply:
you cannot reuse because of the prank right

Reply:
yeah that'll be the small modification the deploy script would need to make rn, otherwise i can move the prank outside the function so the script can re-use and start a broadcast instead (i initially did but it become tricky for this because we had multiple code paths to chose the pranker (proxyAdmin owner is contract, proxyAdminAddress if eoa)
regardless these upgrades won't be performed by an eoa, but rather most a multisig so the script anyway generates calldata and the tests mock

Reply:
I think for tests we could do simple and just prank the proxy admin (no matter if contract or not), but no strong opinion